### DISCUSSION TITLE
Activity Browser in WASM

### DISCUSSION THREAD

**Discussion by michaelweinold** [brightway-lca/brightway-live#47] [Strategy and Planning]
> [!NOTE]
> Following discussions with @bsteubing, I understand that the primary issue for new users of the Activity Browser is the interaction with the Conda/Python ecosystem. It would be interesting to explore the possibility of _"ActivityBrowser in the Browser"_. This discussion is intended for discussion and knowledge-transfer related to this effort. While @michaelweinold won't have time to contribute, he is happy to share the WASM knowledge he has gathered so far with Brightway Live.

<img width="750" alt="Screenshot 2023-09-30 at 10 01 12" src="https://github.com/brightway-lca/brightway-live/assets/23102087/62170953-bda5-4037-a053-7129fbcd0dd3">

The ActivityBrowser currently uses [PySide](https://en.wikipedia.org/wiki/PySide), a Python binding for the cross-platform tool [Qt](https://en.wikipedia.org/wiki/Qt_(software)) to provide a GUI for underlying ActivityBrowser/Brightway functionality.

After some investigation, it seems that there are two viable options to achieve a comparable user experience:

## Option 1: GUI in WASM

The idea of having a complex GUI of a Python package run in the browser, with the Python package also running in the browser is intriguing. There are efforts in nearly every Python GUI framework to make this happen. It is my judgement that even the most advanced efforts (Qt, see below) will take a 1-2 more years until they are production ready.

### Qt

There are efforts to make Qt work with WASM. This [includes the original author of PySide](https://github.com/pyodide/pyodide/issues/1933#issuecomment-1384316851) (=Qt for Python). 

- [Qt Documentation: Qt for WebAssembly](https://doc.qt.io/qt-5/wasm.html)
- [Qt Jira: Qt Python for WebAssembly](https://bugreports.qt.io/browse/PYSIDE-962)
- https://github.com/pyodide/pyodide/issues/1933

> there has been many tasks around this topic, you can see them on the side. Sadly, this is not a simple module that can be plugged into pyodide, the whole build infrastructure is very complex, and requires lots of changes to try to be compatible. It's possible that this will take a long time.

In addition, there are some tutorials that combine Qt with WASM already, but I'm not sure this is what would be applicable here:

- [Create web user interfaces with Qt WebAssembly instead of JavaScript](https://opensource.com/article/20/2/wasm-python-webassembly)

### PySimpleGUI

The developers of [PySimpleGUI](https://www.pysimplegui.org/en/latest/), an alternative to Qt are [also working on WASM support](https://github.com/PySimpleGUI/PySimpleGUI/issues/4944#issuecomment-963210824).

- https://github.com/PySimpleGUI/PySimpleGUI/issues/4944

Conclusion: I am confident that Python (including Python GUIs) in the Browser will be the future of (at the very least teaching) the Python language. Unfortunately, we're not there yet. I recommend not to pursue development efforts to enable a WASM-based GUI for the ActivityBrowser.

## Option 2: Package the ActivityBrowser (`.exe`/`.dmg`)

My knowledge here is limited, but a ChatGPT4/Bing query returned some packages with nice support for Qt-based code:

- Nuitka: A Python compiler that can convert Python code into executable binaries or extension modules. Nuitka supports Qt applications and can handle complex dependencies and imports.
- cx_Freeze: A cross-platform tool that can freeze Python scripts into executable files. cx_Freeze supports Qt applications and can bundle all the required libraries and resources into a single folder or a zip file.
- Shed: A Python-to-C++ compiler that can generate standalone executables from Python code. Shed Skin supports a subset of Python features and can work with Qt applications that use PyQt4 or PySide.
- bbfreeze: A tool that can create standalone executables from Python scripts. bbfreeze supports Qt applications and can include all the dependencies and modules in the executable file.

It seems that this would solve _most_ issues new users face. I don't have the experience to provide an estimate of the additional overhead or development effort that would be involved. Maybe more knowledgable parties (cough @cmutel cough) could chime in.

**Comment by bsteubing** [comment]
thanks so much @michaelweinold - this is a great overview! 

Option 2 should, in principle, be the easier solution, but we have had issues packaging the AB in the past (we have tried it several times over the years and sometimes it worked, we had even a Windows installer for it, but it is never simple... e.g. right now using [pyinstaller](https://pyinstaller.org/en/stable/) produces an .exe file, but then the latter fails as it cannot find a module 'openpyxl.cell._writer', which is apparently used in bw2io... (I didn't go further than that)

**Comment by michaelweinold** [comment]
This might also be interesting: https://wasmer.io
